WP on AWS EKS results in CrashLoopBackoff

**Describe the bug**
I've installed WP on AWS EKS in a dedicated namespace called `waypoint`

The server comes up fine and I'm able to reach it in browser but the runner immediately results in a CrashLoopBackoff which seems to be caused by a SIGSEGV in the runner agent code as indicated by the following log taken from a runner pod:

```
2021-04-30T17:17:43.879Z [INFO] waypoint.runner.agent: initializing the runner
2021-04-30T17:17:43.879Z [INFO] waypoint.runner.agent: starting runner
2021-04-30T17:17:43.879Z [DEBUG] waypoint.runner.agent.runner: registering runner
2021-04-30T17:17:43.879Z [TRACE] waypoint.runner.agent.runner: runner connected, waiting for initial config
2021-04-30T17:17:43.880Z [INFO] waypoint.runner.agent.runner: runner registered with server
2021-04-30T17:17:43.880Z [DEBUG] waypoint.runner.agent.runner: opening job stream
2021-04-30T17:17:43.881Z [TRACE] waypoint.runner.agent.runner: sending job request
2021-04-30T17:17:43.881Z [INFO] waypoint.runner.agent.runner: waiting for job assignment
2021-04-30T17:18:43.426Z [ERROR] waypoint.runner.agent: error running job: err="rpc error: code = Unavailable desc = transport is closing"
2021-04-30T17:18:43.426Z [WARN] waypoint.runner.agent: server unavailable, sleeping before retry
2021-04-30T17:18:43.426Z [ERROR] waypoint.runner.agent.runner.config_recv: error receiving configuration, exiting: err="rpc error: code = Unavailable desc = transport is closing"
2021-04-30T17:18:43.426Z [TRACE] waypoint.runner.agent.runner.config_recv: exiting receive goroutine
2021-04-30T17:18:45.426Z [DEBUG] waypoint.runner.agent.runner: opening job stream
2021-04-30T17:18:45.426Z [TRACE] waypoint.runner.agent.runner: sending job request
2021-04-30T17:18:45.426Z [INFO] waypoint.runner.agent.runner: waiting for job assignment
2021-04-30T17:18:45.428Z [ERROR] waypoint.runner.agent: error running job: err="rpc error: code = NotFound desc = runner ID not found"
2021-04-30T17:18:45.428Z [ERROR] waypoint.runner.agent: runner unexpectedly deregistered, exiting
2021-04-30T17:18:45.428Z [INFO] waypoint.runner.agent: quit request received, gracefully stopping runner
2021-04-30T17:18:45.428Z [WARN] waypoint.runner.agent: error accepting liveness connection: %s: EXTRA_VALUE_AT_END="accept tcp [::]:1234: use of closed network connection"
2021-04-30T17:18:45.428Z [TRACE] waypoint: stopping signal listeners and cancelling the context
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x2626963]

goroutine 65 [running]:
github.com/hashicorp/waypoint/internal/cli.(*RunnerAgentCommand).Run.func1(0x3256190, 0xc00015a7b0, 0x32941b8, 0xc0007c3500)
/tmp/wp-src/internal/cli/runner_agent.go:142 +0x23
created by github.com/hashicorp/waypoint/internal/cli.(*RunnerAgentCommand).Run
/tmp/wp-src/internal/cli/runner_agent.go:132 +0xe65
```


**Steps to Reproduce**

WP CLI install command: `wp install --platform=kubernetes -accept-tos -k8s-namespace=waypoint`


**Expected behavior**
WP server and runner start without CrashLoopBackoff

**Waypoint Platform Versions**
Additional version and platform information to help triage the issue if
applicable:

* Waypoint CLI Version: 0.3.1
* Waypoint Server Platform and Version: kubernetes
* Waypoint Plugin: n/a

* AWS region: us-west-2
* AWS instance type: m5.large
* Kubernetes version: 1.17
* EKS platform version: eks.7 

**Additional context**
Add any other context about the problem here.

@evanphx I'm getting the same issue with EKS with Kubernetes version 1.19 and Waypoint version 0.4.0. Here are the pod's logs:

```
$ kubectl logs waypoint-runner-785855d544-sgz4k
2021-06-28T19:36:38.790Z [INFO]  waypoint: waypoint version: full_string="v0.4.0 (3b3dd831+CHANGES)" version=v0.4.0 prerelease= metadata= revision=3b3dd831+CHANGES
2021-06-28T19:36:38.790Z [TRACE] waypoint: starting interrupt listener for context cancellation
2021-06-28T19:36:38.792Z [DEBUG] waypoint: home configuration directory: path=/home/waypoint/.config/waypoint
2021-06-28T19:36:38.792Z [TRACE] waypoint: no API client provided, initializing connection if possible
2021-06-28T19:36:38.792Z [INFO]  waypoint.server: attempting to source credentials and connect
2021-06-28T19:36:38.794Z [TRACE] waypoint: interrupt listener goroutine started
2021-06-28T19:36:38.831Z [DEBUG] waypoint.server: connection established with sourced credentials
2021-06-28T19:36:38.831Z [TRACE] waypoint: requesting version info from server
2021-06-28T19:36:38.834Z [INFO]  waypoint: server version info: version=v0.4.0 api_min=1 api_current=1 entrypoint_min=1 entrypoint_current=1
2021-06-28T19:36:38.834Z [INFO]  waypoint: negotiated api version: version=1
2021-06-28T19:36:38.834Z [INFO]  waypoint.runner.agent: sourcing credentials and connecting to the Waypoint server
2021-06-28T19:36:38.860Z [INFO]  waypoint.runner.agent: initializing the runner

» Runner configuration:
  Server address: xxxxxxxxxxxxxxxx.us-west-2.elb.amazonaws.com:9701


» Runner logs:

2021-06-28T19:36:38.860Z [INFO]  waypoint.runner.agent: starting runner
2021-06-28T19:36:38.860Z [DEBUG] waypoint.runner.agent.runner: registering runner
2021-06-28T19:36:38.861Z [TRACE] waypoint.runner.agent.runner: runner connected, waiting for initial config
2021-06-28T19:36:38.862Z [INFO]  waypoint.runner.agent.runner: runner registered with server
2021-06-28T19:36:38.862Z [DEBUG] waypoint.runner.agent.runner: opening job stream
2021-06-28T19:36:38.863Z [TRACE] waypoint.runner.agent.runner: sending job request
2021-06-28T19:36:38.863Z [INFO]  waypoint.runner.agent.runner: waiting for job assignment
2021-06-28T19:37:38.169Z [ERROR] waypoint.runner.agent: error running job: err="rpc error: code = Unavailable desc = transport is closing"
2021-06-28T19:37:38.169Z [WARN]  waypoint.runner.agent: server unavailable, sleeping before retry
2021-06-28T19:37:38.169Z [ERROR] waypoint.runner.agent.runner.config_recv: error receiving configuration, exiting: err="rpc error: code = Unavailable desc = transport is closing"
2021-06-28T19:37:38.169Z [TRACE] waypoint.runner.agent.runner.config_recv: exiting receive goroutine
2021-06-28T19:37:40.169Z [DEBUG] waypoint.runner.agent.runner: opening job stream
2021-06-28T19:37:40.169Z [TRACE] waypoint.runner.agent.runner: sending job request
2021-06-28T19:37:40.169Z [INFO]  waypoint.runner.agent.runner: waiting for job assignment
2021-06-28T19:37:40.171Z [ERROR] waypoint.runner.agent: error running job: err="rpc error: code = NotFound desc = runner ID not found"
2021-06-28T19:37:40.171Z [ERROR] waypoint.runner.agent: runner unexpectedly deregistered, exiting
2021-06-28T19:37:40.171Z [INFO]  waypoint.runner.agent: quit request received, gracefully stopping runner
2021-06-28T19:37:40.171Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-28T19:37:40.171Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-28T19:37:40.171Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-28T19:37:40.171Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-28T19:37:40.171Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-28T19:37:40.171Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-28T19:37:40.171Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-28T19:37:40.171Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-28T19:37:40.171Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-28T19:37:40.171Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-28T19:37:40.171Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-28T19:37:40.171Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-28T19:37:40.171Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-28T19:37:40.171Z [TRACE] waypoint: stopping signal listeners and cancelling the context
```

--- a/.changelog/1735.txt
+++ b/.changelog/1735.txt
@@ -1,0 +1,4 @@
+```release-note:bug
+server/client: configure keepalive properties to RPC connections to persist
+connections even after inactivity
+```

--- a/internal/server/grpc.go
+++ b/internal/server/grpc.go
@@ -6,6 +6,7 @@ import (
 	"github.com/golang/protobuf/ptypes/empty"
 	"github.com/oklog/run"
 	"google.golang.org/grpc"
+	"google.golang.org/grpc/keepalive"
 	"google.golang.org/grpc/reflection"
 
 	pb "github.com/hashicorp/waypoint/internal/server/gen"
@@ -37,6 +38,15 @@ func grpcInit(group *run.Group, opts *options) error {
 			// Protocol version negotiation
 			versionStreamInterceptor(resp.Info),
 		),
+		grpc.KeepaliveEnforcementPolicy(
+			keepalive.EnforcementPolicy{
+				// connections need to wait at least 20s before sending a
+				// keepalive ping
+				MinTime: 20 * time.Second,
+				// allow runners to send keeplive pings even if there are no
+				// active RCP streams.
+				PermitWithoutStream: true,
+			}),
 	)
 
 	if opts.AuthChecker != nil {

--- a/internal/serverclient/client.go
+++ b/internal/serverclient/client.go
@@ -10,6 +10,7 @@ import (
 
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/keepalive"
 
 	"github.com/hashicorp/waypoint/internal/clicontext"
 	"github.com/hashicorp/waypoint/internal/protocolversion"
@@ -56,6 +57,13 @@ func Connect(ctx context.Context, opts ...ConnectOption) (*grpc.ClientConn, erro
 		grpc.WithBlock(),
 		grpc.WithUnaryInterceptor(protocolversion.UnaryClientInterceptor(protocolversion.Current())),
 		grpc.WithStreamInterceptor(protocolversion.StreamClientInterceptor(protocolversion.Current())),
+		grpc.WithKeepaliveParams(
+			keepalive.ClientParameters{
+				// ping after this amount of time of inactivity
+				Time: 30 * time.Second,
+				// send keepalive pings even if there is no active streams
+				PermitWithoutStream: true,
+			}),
 	}
 
 	if !cfg.Tls {

